prow/plugins/approve/approve.go

```diff
@@ -0,0 +1,374 @@
+		return nil
+	}
+
+	return handle(
+		log,
+		ghc,
+		repo,
+		optionsForRepo(config, ce.Repo.Owner.Login, ce.Repo.Name),
+		&state{
+			org:       ce.Repo.Owner.Login,
+			repo:      ce.Repo.Name,
+			number:    ce.Number,
+			body:      ce.IssueBody,
+			author:    ce.IssueAuthor.Login,
+			assignees: ce.Assignees,
+			htmlURL:   ce.IssueHTMLURL,
+		},
+	)
+}
+
+func handlePullRequestEvent(pc plugins.PluginClient, pre github.PullRequestEvent) error {
+	ro, err := pc.OwnersClient.LoadRepoOwners(pre.Repo.Owner.Login, pre.Repo.Name)
+	if err != nil {
+		return err
+	}
+	return handlePullRequest(pc.Logger, pc.GitHubClient, ro, pc.PluginConfig, &pre)
+}
+
+func handlePullRequest(log *logrus.Entry, ghc githubClient, repo approvers.RepoInterface, config *plugins.Configuration, pre *github.PullRequestEvent) error {
+	if pre.Action != github.PullRequestActionOpened &&
+		pre.Action != github.PullRequestActionReopened &&
+		pre.Action != github.PullRequestActionSynchronize {
+		return nil
+	}
+
+	return handle(
+		log,
+		ghc,
+		repo,
+		optionsForRepo(config, pre.Repo.Owner.Login, pre.Repo.Name),
+		&state{
+			org:       pre.Repo.Owner.Login,
+			repo:      pre.Repo.Name,
+			number:    pre.Number,
+			body:      pre.PullRequest.Body,
+			author:    pre.PullRequest.User.Login,
+			assignees: pre.PullRequest.Assignees,
+			htmlURL:   pre.PullRequest.HTMLURL,
+		},
+	)
+}
+
+// Returns associated issue, or 0 if it can't find any.
+// This is really simple, and could be improved later.
+func findAssociatedIssue(body string) int {
+	match := AssociatedIssueRegex.FindStringSubmatch(body)
+	if len(match) == 0 {
+		return 0
+	}
+	v, err := strconv.Atoi(match[1])
+	if err != nil {
+		return 0
+	}
+	return v
+}
+
+// handle is the workhorse the will actually make updates to the PR.
+// The algorithm goes as:
+// - Initially, we build an approverSet
+//   - Go through all comments in order of creation.
+//		 - (Issue/PR comments, PR review comments, and PR review bodies are considered as comments)
+//	 - If anyone said "/approve" or "/lgtm", add them to approverSet.
+// - Then, for each file, we see if any approver of this file is in approverSet and keep track of files without approval
+//   - An approver of a file is defined as:
+//     - Someone listed as an "approver" in an OWNERS file in the files directory OR
+//     - in one of the file's parent directorie
+// - Iff all files have been approved, the bot will add the "approved" label.
+// - Iff a cancel command is found, that reviewer will be removed from the approverSet
+// 	and the munger will remove the approved label if it has been applied
+func handle(log *logrus.Entry, ghc githubClient, repo approvers.RepoInterface, opts *plugins.Approve, pr *state) error {
+	fetchErr := func(context string, err error) error {
+		return fmt.Errorf("failed to get %s for %s/%s#%d: %v", context, pr.org, pr.repo, pr.number, err)
+	}
+
+	filenames := []string{}
```

Comment:
nit: Move this next to where it's populated and define it as `var filenames []string`